File: packages/provide-uterm/src/provide/uterm/vnc/rfb_filter.py
#
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
# SPDX-License-Identifier: AGPL-3.0-or-later
#
"""RFB client→server input filter (Go vnc.filterRFBInput parity).

Used for human VNC relay: pass-through handshake + non-input messages;
gate KeyEvent / PointerEvent / ClientCutText on a CanInject callback.
Nil / missing inject fn fails closed (drops inject messages).
"""

from __future__ import annotations

import contextlib
import struct
from collections.abc import Callable
from contextlib import AbstractContextManager
from typing import Any, BinaryIO

# Client message types (RFB 3.8)
_SET_PIXEL_FORMAT = 0
_SET_ENCODINGS = 2
_FRAMEBUFFER_UPDATE_REQUEST = 3
_KEY_EVENT = 4
_POINTER_EVENT = 5
_CLIENT_CUT_TEXT = 6

MAX_CUT_TEXT = 1 << 20  # 1 MiB

CanInjectFn = Callable[[str, str, str, str], bool]
"""(session_id, lease_id, principal_id, principal_role) -> allowed."""


def filter_rfb_client_input(
    dst: BinaryIO,
    src: BinaryIO,
    *,
    can_inject: CanInjectFn | None,
    session_id: str,
    lease_id: str,
    principal_id: str,
    principal_role: str,
    dst_lock: AbstractContextManager[Any] | None = None,
    on_client_ready: Callable[[], None] | None = None,
) -> None:
    """Copy RFB client messages from *src* to *dst*, gating inject types.

    Each RFB message is written to *dst* in a single ``write`` guarded by
    *dst_lock* (a no-op context when ``None``). This keeps writes atomic so a
    concurrent writer of *dst* — e.g. the update-driver thread in
    :func:`run_human_relay_streams` that injects periodic
    ``FramebufferUpdateRequest`` frames — never interleaves mid-message.

    *on_client_ready*, if given, is invoked once the client's FIRST
    ``FramebufferUpdateRequest`` has been forwarded. By then the client's
    ``SetPixelFormat`` + ``SetEncodings`` (which precede that request) are already
    upstream, so a driver that starts injecting requests won't race ahead of the
    client's pixel format — otherwise the server may answer the driver's request
    in its native format and the client renders those frames with swapped colours.

    Raises ``ValueError`` on unsupported security type or unknown message type.
    Raises ``EOFError`` on short read.
    """
    guard: AbstractContextManager[Any] = dst_lock if dst_lock is not None else contextlib.nullcontext()

    def emit(data: bytes) -> None:
        with guard:
            dst.write(data)

    # 1. ProtocolVersion (12 bytes)
    emit(_read_exact(src, 12))

    # 2. Security type (1 byte) — only None (1)
    sec = _read_exact(src, 1)
    if sec[0] != 1:
        raise ValueError(f"unsupported security type {sec[0]}")
    emit(sec)

    # 3. ClientInit (1 byte)
    emit(_read_exact(src, 1))

    client_ready_fired = False
    while True:
        try:
            msg_type = _read_exact(src, 1)
        except EOFError:
            return
        t = msg_type[0]
        if t == _SET_PIXEL_FORMAT:
            emit(msg_type + _read_exact(src, 19))
        elif t == _SET_ENCODINGS:
            header = _read_exact(src, 3)
            num = struct.unpack("!H", header[1:3])[0]
            body = _read_exact(src, num * 4) if num > 0 else b""
            emit(msg_type + header + body)
        elif t == _FRAMEBUFFER_UPDATE_REQUEST:
            emit(msg_type + _read_exact(src, 9))
            if on_client_ready is not None and not client_ready_fired:
                client_ready_fired = True
                on_client_ready()
        elif t == _KEY_EVENT:
            payload = _read_exact(src, 7)
            if _allowed(can_inject, session_id, lease_id, principal_id, principal_role):
                emit(msg_type + payload)
        elif t == _POINTER_EVENT:
            payload = _read_exact(src, 5)
            if _allowed(can_inject, session_id, lease_id, principal_id, principal_role):
                emit(msg_type + payload)
        elif t == _CLIENT_CUT_TEXT:
            header = _read_exact(src, 7)
            length = struct.unpack("!I", header[3:7])[0]
            # RFB extended clipboard sets the high bit of the length field; the
            # remaining 31 bits are the real payload size (noVNC uses this).
            payload_len = length & 0x7FFFFFFF
            if payload_len > MAX_CUT_TEXT:
                # Oversized / hostile cut-text: drain what we can and drop the
                # message so the display session stays up (raising would tear
                # down the human relay and black the framebuffer).
                remaining = payload_len
                while remaining > 0:
                    chunk = src.read(min(remaining, 65_536))
                    if not chunk:
                        return
                    remaining -= len(chunk)
                continue
            payload = _read_exact(src, payload_len) if payload_len else b""
            if _allowed(can_inject, session_id, lease_id, principal_id, principal_role):
                emit(msg_type + header + payload)
        else:
            raise ValueError(f"unknown RFB client message type: {t}")


def _allowed(
    can_inject: CanInjectFn | None,
    session_id: str,
    lease_id: str,
    principal_id: str,
    principal_role: str,
) -> bool:
    if can_inject is None:
        return False
    return bool(can_inject(session_id, lease_id, principal_id, principal_role))


def _read_exact(src: BinaryIO, n: int) -> bytes:
    buf = src.read(n)
    if buf is None or len(buf) < n:
        raise EOFError(f"short read: want {n}, got {0 if buf is None else len(buf)}")
    return buf

File: packages/provide-uterm-ts/src/vnc/rfb-filter.ts
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

/**
 * What a viewer is allowed to send a graphical session.
 *
 * Port of `provide.uterm.vnc.rfb_filter`, which is itself parity with the Go
 * port's `filterRFBInput`.
 *
 * Everything that only *reads* the screen passes through. The three messages
 * that act on it — keystrokes, pointer movement, clipboard writes — are gated
 * on a permission check. This is the boundary that stops somebody watching a
 * session from typing into it.
 */

/** The client messages this filter understands. */
const SET_PIXEL_FORMAT = 0;
const SET_ENCODINGS = 2;
const FRAMEBUFFER_UPDATE_REQUEST = 3;
const KEY_EVENT = 4;
const POINTER_EVENT = 5;
const CLIENT_CUT_TEXT = 6;

/** How much clipboard a viewer may paste at once. */
export const MAX_CUT_TEXT = 1 << 20;

/** The only security type the relay implements. */
const SECURITY_NONE = 1;

/** How many bytes each fixed-size message carries after its type byte. */
const PIXEL_FORMAT_BYTES = 19;
const UPDATE_REQUEST_BYTES = 9;
const KEY_EVENT_BYTES = 7;
const POINTER_EVENT_BYTES = 5;
const CUT_TEXT_HEADER_BYTES = 7;
const ENCODINGS_HEADER_BYTES = 3;

/** How much is drained at a time when discarding an oversized message. */
const DRAIN_CHUNK = 65_536;

/** RFB's extended clipboard sets the top bit; the rest is the real length. */
const EXTENDED_CLIPBOARD_MASK = 0x7fffffff;

/** A stream ended in the middle of a message. */
export class RfbShortReadError extends Error {}

/** A stream said something this filter does not implement. */
export class RfbProtocolError extends Error {}

/** Whether this viewer may act on the session, rather than only watch it. */
export type CanInject = (sessionId: string, leaseId: string, principalId: string, principalRole: string) => boolean;

/** Where the client's bytes come from. */
export interface ByteSource {
  /** Up to `size` bytes, or fewer at the end of the stream. */
  read(size: number): Uint8Array;
}

/** Where the filtered bytes go. */
export interface ByteSink {
  write(data: Uint8Array): void;
}

/** Who is sending, and what to do about it. */
export interface RfbFilterOptions {
  /** Omitted means nobody may inject — see {@link filterRfbClientInput}. */
  canInject?: CanInject | undefined;
  sessionId: string;
  leaseId: string;
  principalId: string;
  principalRole: string;
  /** Called once the client's first update request has gone upstream. */
  onClientReady?: (() => void) | undefined;
}

/** Read exactly `size` bytes, or fail. */
function readExact(src: ByteSource, size: number): Uint8Array {
  const buffer = src.read(size);
  if (buffer.length < size) {
    throw new RfbShortReadError(`short read: want ${size}, got ${buffer.length}`);
  }
  return buffer;
}

/** Join byte runs into one write, so a message reaches the sink whole. */
function joined(...parts: readonly Uint8Array[]): Uint8Array {
  const total = parts.reduce((sum, part) => sum + part.length, 0);
  const out = new Uint8Array(total);
  let offset = 0;
  for (const part of parts) {
    out.set(part, offset);
    offset += part.length;
  }
  return out;
}

/**
 * Whether an injecting message may be forwarded.
 *
 * No callback means no: a relay wired up without one would otherwise forward
 * keystrokes from anybody, and failing closed is the only safe reading of
 * "nobody said who may send them".
 */
function allowed(options: RfbFilterOptions): boolean {
  return (
    options.canInject !== undefined &&
    options.canInject(options.sessionId, options.leaseId, options.principalId, options.principalRole)
  );
}

/**
 * Copy a client's RFB messages, gating the ones that act on the session.
 *
 * A refused message is dropped rather than refused: the stream is a byte
 * protocol with no room for an error, so a keystroke a viewer may not send is
 * simply not forwarded — the session stays up and the viewer sees nothing
 * happen.
 *
 * Each message is written in one call, so a concurrent writer of the same sink
 * — the driver injecting periodic update requests — cannot interleave halfway
 * through one.
 *
 * @throws {RfbProtocolError} On a security type or message type this does not
 *   implement.
 * @throws {RfbShortReadError} When the stream ends mid-message.
 */
export function filterRfbClientInput(dst: ByteSink, src: ByteSource, options: RfbFilterOptions): void {
  // The handshake passes through as it is: a version, a security type, and
  // the client's shared-desktop flag.
  dst.write(readExact(src, 12));

  const security = readExact(src, 1);
  if (security[0] !== SECURITY_NONE) {
    throw new RfbProtocolError(`unsupported security type ${security[0]}`);
  }
  dst.write(security);
  dst.write(readExact(src, 1));

  let clientReadyFired = false;
  for (;;) {
    const header = src.read(1);
    if (header.length < 1) {
      // The client hung up between messages, which is how a session ends.
      return;
    }

    switch (header[0]) {
      case SET_PIXEL_FORMAT:
        dst.write(joined(header, readExact(src, PIXEL_FORMAT_BYTES)));
        break;

      case SET_ENCODINGS: {
        const encodings = readExact(src, ENCODINGS_HEADER_BYTES);
        const count = (encodings[1] as number) * 256 + (encodings[2] as number);
        const body = count > 0 ? readExact(src, count * 4) : new Uint8Array(0);
        dst.write(joined(header, encodings, body));
        break;
      }

      case FRAMEBUFFER_UPDATE_REQUEST:
        dst.write(joined(header, readExact(src, UPDATE_REQUEST_BYTES)));
        // Announced only once, and only after the request is upstream: the
        // client's pixel format and encodings precede it, so a driver that
        // starts injecting requests before this would have the server answer
        // in its own format and the client render those frames with swapped
        // colours.
        if (options.onClientReady !== undefined && !clientReadyFired) {
          clientReadyFired = true;
          options.onClientReady();
        }
        break;

      case KEY_EVENT: {
        const payload = readExact(src, KEY_EVENT_BYTES);
        if (allowed(options)) {
          dst.write(joined(header, payload));
        }
        break;
      }

      case POINTER_EVENT: {
        const payload = readExact(src, POINTER_EVENT_BYTES);
        if (allowed(options)) {
          dst.write(joined(header, payload));
        }
        break;
      }

      case CLIENT_CUT_TEXT: {
        const cutHeader = readExact(src, CUT_TEXT_HEADER_BYTES);
        const declared =
          (cutHeader[3] as number) * 0x1000000 +
          (cutHeader[4] as number) * 0x10000 +
          (cutHeader[5] as number) * 0x100 +
          (cutHeader[6] as number);
        // The extended clipboard sets the top bit; reading the field whole
        // would make every extended write look like two gigabytes.
        const length = declared & EXTENDED_CLIPBOARD_MASK;

        if (length > MAX_CUT_TEXT) {
          // Drained and dropped rather than refused: raising would tear down
          // the relay and black the framebuffer for everyone watching, which
          // is a worse answer to one hostile message.
          //
          // The drain trusts the declared length, so a client that declares
          // more than it sends has its whole remaining stream consumed and
          // the session's input ends here. That is the reference's behaviour.
          let remaining = length;
          while (remaining > 0) {
            const chunk = src.read(Math.min(remaining, DRAIN_CHUNK));
            if (chunk.length === 0) {
              return;
            }
            remaining -= chunk.length;
          }
          break;
        }

        const payload = length > 0 ? readExact(src, length) : new Uint8Array(0);
        if (allowed(options)) {
          dst.write(joined(header, cutHeader, payload));
        }
        break;
      }

      default:
        throw new RfbProtocolError(`unknown RFB client message type: ${header[0]}`);
    }
  }
}
